File: src/day04.hs
module Main where

import Data.Word (Word8 )

import Data.Char (ord)
import Data.List (isPrefixOf)

import Data.Digest.Pure.MD5 (md5 )
import Data.ByteString.Lazy (ByteString, pack)




part1 :: String -> Int
part1 s =  head $ filter ( valid1 . (s++) . show    ) [ 1..]

part2 :: String -> Int
part2 s =  head $ filter ( valid2 . (s++) . show    ) [ 1..]


valid1 :: String -> Bool
valid1 = (isPrefixOf "00000") . parse

valid2 :: String -> Bool
valid2 = (isPrefixOf "000000") . parse


parse :: String -> String
parse = show . md5 . pack . map ( fromIntegral . ord )

main :: IO ()
main = do 
          file <- readFile ("../data/04.txt")
          putStrLn ( "p1: " ++ (show $ part1 file ) )
          putStrLn ( "p2: " ++ (show $ part2 file ) )

File: app/day04.py
from itertools import count
from functools import partial
from hashlib import md5


def digest(s):
    return str(md5(bytes(s, encoding="ascii")).hexdigest())


def part1(key):
    return next(
        filter(lambda k: (digest(key+k)).startswith("00000"), map(str, count(1)))
    )

def part2(key):
    return next(
        filter(lambda k: (digest(key+k)).startswith("000000"), map(str, count(1)))
    )



def main():
    with open("../data/04.txt") as f:
        w = f.read()
        print(part1(w))
        print(part2(w))


if __name__ == "__main__":
    main()
